tools/utils/cancelation.js

```diff
@@ -0,0 +1,41 @@
+class CancelError extends Error {}
+
+export class CancelToken {
+  constructor(exec) {
+    let requested = false;
+    let message;
+
+    exec(function cancel(reason = "cancelation requested") {
+      requested = true;
+      message = String(reason);
+    });
+
+    this.throwIfRequested = function () {
+      // It's important to introduce a gap that's long enough for file
+      // change notifications to fire, and only setTimeout seems to work.
+      new Promise(resolve => setTimeout(resolve, 0)).await();
+      if (requested) {
+        throw new CancelError(message);
```

Comment:
This `CancelError` can be caught/ignored, of course, but sometimes that power is useful, and exceptions are still the best tool for canceling work in JavaScript.